Weather App 5 day forecast

clintonmedbery opened the pull request:
- [ ]  As a user, I would like to search for the current weather in my area
- [ ]   As a user, I would like to see the 5-day forecast
- [ ]   As a user, I would like to share the URL of the weather in my area and see the results
- [ ]   As a user, I would like to toggle between celsius and fahrenheit

First off apologies as this does not target a main branch. I know not to work off master but for this exercise I did so I am merging into 'main'.

![image](https://user-images.githubusercontent.com/7118459/104986581-5300b100-59e1-11eb-8f49-75c971d85449.png)

Added pre-commit hooks to prettify code and run linter on staged files.

Switched to weatherbit API. They provided some nice icons (icons provided did not include all weather like snow) and allowed me to search by city for forecasts.

Didn't have time for testing, so I will explain my usual approach.

I try to always practice TDD and usually start a story by writing backend tests. I prefer integration over unit testing. Often times unit tests miss how the parts all work together, but I can see the need for complex functions. I also think it is important to test for security and authorization. For example, can User A access User B's data? There needs to be tests to ensure data can only be accessed by the right parties.

On the front end you can use React Testing Library or jest to test the functionality of components. 

I also have experience using tools like SonarQube to track code coverage.

clintonmedbery commented on package.json:
```diff
@@ -21,5 +34,20 @@
     "not op_mini all"
-  ]
+  ],
+  "husky": {
```
It's important to get some rules set up early before the code base becomes the wild west.

clintonmedbery commented on src/components/app.jsx:
```diff
@@ -1,9 +1,15 @@
+import '../App.css'
+import ForecastPage from './modules/forecast-page/forecast-page.component'
+import 'antd/dist/antd.css'
+import { BrowserRouter as Router, Route } from 'react-router-dom'
```
There is no routing here but we are using this to share URLs. Also no redux or higher order Context API here since the state is simple.

clintonmedbery commented on src/components/business-components/weather-card/weather-card.component.js:
```diff
@@ -0,0 +1,38 @@
+}
+
+const WeatherCard = ({ iconName, description, date, temperature, unit }) => {
+  const dayOfWeek = dayjs(date).format('dddd')
```
dayjs for to get day names. In the past I am have used moment but it's been deprecated.

clintonmedbery commented on src/components/modules/forecast-page/forecast-page.component.js:
```diff
@@ -0,0 +1,135 @@
+
+  return (
+    <>
+      <div style={{ margin: '2em' }}>
```
This page header could be moved to it's own component, but it's simple enough and the project scope is small enough I left it here.

clintonmedbery commented on src/components/modules/forecast-page/forecast-utilities.js:
```diff
@@ -0,0 +1,10 @@
+export const formatWeatherData = (weatherData) => {
```
Probably could have left this in the component, but typically I try to move some logic to local utility functions to keep the component tidy.

clintonmedbery commented on src/components/ui-components/autosuggest.component.js:
```diff
@@ -0,0 +1,79 @@
+import React, { useState, useMemo } from 'react'
```
I like keeping more pure UI Components separate from components with more specific business logic. I notice a lot of devs end up putting a lot of business logic in lower components making them hard or impossible to reuse. Or they have a ton of flags in the components. This keeps things clean.

clintonmedbery commented on src/custom-hooks/use-debounce.js:
```diff
@@ -0,0 +1,20 @@
+import { useState, useEffect } from 'react'
```
I included this function to not only show I know hooks well enough to make them, but also because we don't want to be making the cities call every stroke.

clintonmedbery commented on src/services/location-service.js:
```diff
@@ -0,0 +1,17 @@
+export const fetchCities = (cityInput) => {
```
Yes this is fake, we could get a more exhaustive list or make a service. We should also include state info so we aren't getting back the wrong Charleston etc. This works for this exercise.

jlaurelli commented:
Hi Clinton! I'm James, one of the engineers here at Contently.

Your approach is pretty interesting. I like that you already went through and commented on some of your architectural decisions ahead of time.

What would be some things you'd do if this were part of a larger app? You mentioned about not using state management here due to project scope (which is fine), but how would you envision handling an issue like this if this project were part of a wider one?

Additionally, what would you foresee as ways of handling caching in a similar scenario?

Are there any other changes you would also make in this situation?

I'm just picking your brain a bit. Just high level approaches would be fine.

clintonmedbery commented:
Hey James!

> What would be some things you'd do if this were part of a larger app?

I would strongly consider breaking out base components into a common/shared npm package. At my last two jobs we have had to do this after the fact in order to share components among libraries and it might be better to do that from the get go. I would also talk to the team about using storybook. I think it makes people think a bit more about the components they are creating and creates value for UI/UX designers. 

Some of the other stuff I have already done, such as installing husky, prettier and lint-staged to help insure code quality. I moved the api calls into services files. I like keeping components clean by putting business logic in utility files. React is very non-opinionated so getting organization down upfront is key.

> You mentioned about not using state management here due to project scope (which is fine), but how would you envision handling an issue like this if this project were part of a wider one?

Well there is redux and Mobx. Lately though I have been pushing for state management using the provided Context API. Redux seems to prefer a lot of boilerplate and at times can be difficult to debug (although the redux rewind feature can help with debugging in some cases). With Context I can create higher order components to wrap child components in, and then access state as needed with useContext. 

Not really sure that there is a right or wrong answer, but I would lean towards Context, since that seems to be where the React team is headed: https://stackoverflow.com/a/49569183/3058839

As a full stack developer, we normally handle caching on the backend for API calls. In this case the API isn't ours, so I would probably use local storage. When we retrieve some data, we could write it to local storage by date and city. If there is info for that date and city we could use that data, and erase any old data from previous days. If there is no current weather data, we make out call, or maybe make the call to the API. If I was to seriously make an app like this, I would probably make a useApi hook that used some caching in it, or follow some of this advice:

https://www.smashingmagazine.com/2020/07/custom-react-hook-fetch-cache-data/
https://medium.com/@brockreece/frontend-caching-strategies-38c57f59e254

> Are there any other changes you would also make in this situation?

I am sure with enough time I could think of more and more. One question that was popping up was WeatherCard specific enough to be a business component or should we make it into a more generic ImageCard that is reusable. A lot of the other changes I addressed in the PR and in this post. I made the card responsive for mobile but I am sure there are other mobile friendly things I could do as well. I am sure I will think of something after I press comment 😂. 

Thanks for taking your time to review and feel free to ask me any additional questions!

jlaurelli commented:
Hi Clinton,

Awesome! Thanks for the response! Very interesting insight.
